src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs

```diff
@@ -0,0 +1,402 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics.EngineV2
+{
+    // TODO: make it to use cache
+    internal partial class DiagnosticIncrementalAnalyzer : BaseDiagnosticIncrementalAnalyzer
+    {
+        public override Task AnalyzeSyntaxAsync(Document document, CancellationToken cancellationToken)
+        {
+            return AnalyzeDocumentForKindAsync(document, AnalysisKind.Syntax, cancellationToken);
+        }
+
+        public override Task AnalyzeDocumentAsync(Document document, SyntaxNode bodyOpt, CancellationToken cancellationToken)
+        {
+            return AnalyzeDocumentForKindAsync(document, AnalysisKind.Semantic, cancellationToken);
+        }
+
+        private async Task AnalyzeDocumentForKindAsync(Document document, AnalysisKind kind, CancellationToken cancellationToken)
+        {
+            try
+            {
+                if (!AnalysisEnabled(document))
+                {
+                    // to reduce allocations, here, we don't clear existing diagnostics since it is dealt by other entry point such as
+                    // DocumentReset or DocumentClosed.
+                    return;
+                }
+
+                var stateSets = _stateManager.GetOrUpdateStateSets(document.Project);
+                var analyzerDriverOpt = await _compilationManager.GetAnalyzerDriverAsync(document.Project, stateSets, cancellationToken).ConfigureAwait(false);
+
+                foreach (var stateSet in stateSets)
```

Comment:
I didn't do that to make sure stateSets is immutable. (since I cache CompilationWithAnalyzer based only on Project snapshot). can IsAnalyzerSuppressed be changed with same project snapshot?